exec/java-exec/src/main/java/org/apache/drill/exec/store/log/LogRecordReader.java

```diff
@@ -0,0 +1,623 @@
+      return new VarCharDefn(name, patternIndex);
+    }
+
+    MinorType type = MinorType.valueOf(typeName);
+    switch (type) {
+      case VARCHAR:
+        return new VarCharDefn(name, patternIndex);
+      case INT:
+      case SMALLINT:
+      case BIGINT:
+        return new IntDefn(name, patternIndex);
+      case FLOAT8:
+      case FLOAT4:
+      case DECIMAL9:
+      case DECIMAL18:
+        return new DoubleDefn(name, patternIndex);
+      case DATE:
+        return new DateDefn(name, patternIndex, formatConfig.getDateFormat());
+      case TIMESTAMP:
+        return new TimeStampDefn(name, patternIndex, formatConfig.getDateFormat());
+      case TIME:
+        return new TimeDefn(name, patternIndex, formatConfig.getTimeFormat());
+      default:
+        throw UserException
+            .validationError()
+            .message("Undefined column types")
+            .addContext("Position", patternIndex)
+            .addContext("Field name", name)
+            .addContext("Type", typeName)
+            .build(logger);
+    }
+  }
+
+  private void defineVectors() {
+    for (int i = 0; i < columns.length; i++) {
+      try {
+        columns[i].define(outputMutator);
+      } catch (SchemaChangeException e) {
+        throw UserException
+            .systemError(e)
+            .message("Vector creation failed")
+            .build(logger);
+      }
+    }
+  }
+
+  @Override
+  public int next() {
+    rowIndex = 0;
+    while (nextLine()) {
+    }
+    return rowIndex;
+  }
+
+  private boolean nextLine() {
+    String line;
+    try {
+      line = reader.readLine();
+    } catch (IOException e) {
+      throw UserException
+          .dataReadError(e)
+          .message("Error reading file:")
+          .addContext("File", fileWork.getPath())
+          .build(logger);
+    }
+    if (line == null) {
+      return false;
+    }
+    Matcher m = pattern.matcher(line);
+    if (m.matches()) {
+      loadVectors(m);
+      return rowIndex < BATCH_SIZE;
+    }
+    errorCount++;
+    if (errorCount < formatConfig.getMaxErrors()) {
+      logger.warn("Unmatached line: {}", line);
+    } else if( errorCount > formatConfig.getMaxErrors() ){
+      throw UserException.parseError()
+          .message("Too many errors.  Max error threshold exceeded.")
+          .addContext("Line", line)
+          .addContext("Line number", rowIndex)
+          .build(logger);
+    }
+    //If the user asked for the unmatched columns display them
+    if(unmatchedColumnIndex != -1 ) {
+      columns[unmatchedColumnIndex].load(rowIndex, line);
+      rowIndex++;
+      return rowIndex < BATCH_SIZE;
+    }
+
+
+    return true;
+  }
+
+  private void loadVectors(Matcher m) {
+    for (int i = 0; i < columns.length; i++) {
+      if (columns[i].index >= 0) {
+        String value = m.group(columns[i].index + 1);
+        if (value != null) {
+          columns[i].load(rowIndex, value);
+        }
+      } else if( columns[i].index == -1 ){
```

Comment:
Oh, my. In my example, the value of `-1` indicated a column that was projected, but did not exist in the data source and so should always be null (the default value).

Here, the value of `-1` is being used to fill in the entire row. So, if I select `bogus` from a file, the value won't be null, it will be the input row.

Suggestion: this behavior is needed only for the `_raw` column. Simply set a field to the index of the raw column if it exists, as was done earlier for the `_unmatched` column.